### PR TITLE
Add Core Model / Components in ref doc

### PR DESCRIPTION
**Pull request by jgrandja**
Closes gh-671

### CODE REVIEW

File: docs/src/docs/asciidoc/core-model-components.adoc [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+	private final String principalName; <2>
+	private final Set<GrantedAuthority> authorities;    <3>
+
+	...
+
+}
+----
+<1> `registeredClientId`: The ID that uniquely identifies the <<registered-client, RegisteredClient>>.
+<2> `principalName`: The principal name of the resource owner.
+<3> `authorities`: The authorities granted to the client by the resource owner. An authority can represent a scope, claim, permission, role, etc.
```
Comment (Buzzardo) [inline docs/src/docs/asciidoc/core-model-components.adoc:253]: Change "a scope, claim, permission, role, etc." to "a scope, a claim, a permission, a role, and others."

Too many people (according to the studies I've read, about half) misinterpret "etc."

File: docs/src/docs/asciidoc/core-model-components.adoc [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+	...
+
+	return http.build();
+}
+----
+
+[[oauth2-token-customizer]]
+== OAuth2TokenCustomizer
+
+An `OAuth2TokenCustomizer` provides the capability to customize the attributes of an `OAuth2Token`, which are accessible in the provided <<oauth2-token-context, OAuth2TokenContext>>.
```
Comment (Buzzardo) [inline docs/src/docs/asciidoc/core-model-components.adoc:406]: Change "capability" to "ability" (a change that can nearly always be made - "capability" is much overused).

File: docs/src/docs/asciidoc/guides/how-to-jpa.adoc [hunk 3]
```diff
@@ -9,19 +9,19 @@
 [[jpa-getting-started]]
 == Getting Started
 
-In this guide, we will demonstrate how to implement the xref:{docs-dir}/core-components.adoc#core-components[core services] of xref:{docs-dir}/index.adoc#top[Spring Authorization Server] with JPA. The purpose of this guide is to provide a starting point for implementing these services yourself, with the intention that you will make modifications as necessary to suit your needs.
+In this guide, we will demonstrate how to implement the xref:{docs-dir}/core-model-components.adoc#core-model-components[core services] of xref:{docs-dir}/index.adoc#top[Spring Authorization Server] with JPA. The purpose of this guide is to provide a starting point for implementing these services yourself, with the intention that you will make modifications as necessary to suit your needs.
```
Comment (Buzzardo) [inline docs/src/docs/asciidoc/guides/how-to-jpa.adoc:12]: Change to:

This guide shows how to implement the xref:{docs-dir}/core-model-components.adoc#core-model-components[core services] of xref:{docs-dir}/index.adoc#top[Spring Authorization Server] with JPA. The purpose of this guide is to provide a starting point for implementing these services yourself, with the intention that you can make modifications to suit your needs.

File: docs/src/docs/asciidoc/guides/how-to-jpa.adoc [hunk 4]
```diff
@@ -9,19 +9,19 @@
...
 
 [[jpa-define-data-model]]
 == Define the data model
 
-This guide seeks to provide a starting point for the data model and uses the simplest possible structure and data types. To come up with the initial schema, we begin by reviewing the xref:{docs-dir}/core-components.adoc#core-components[domain objects] used by the core services. Please note:
+This guide seeks to provide a starting point for the data model and uses the simplest possible structure and data types. To come up with the initial schema, we begin by reviewing the xref:{docs-dir}/core-model-components.adoc#core-model-components[domain objects] used by the core services. Please note:
```
Comment (Buzzardo) [inline docs/src/docs/asciidoc/guides/how-to-jpa.adoc:17]: Change "seeks to provide" to "provides" and remove "Please note:"

File: docs/src/docs/asciidoc/guides/how-to-jpa.adoc [hunk 5]
```diff
@@ -9,19 +9,19 @@
...
 
 NOTE: Except for token, state, metadata, settings and claims values, we will use the JPA default column length of 255 for all columns. In reality, the length and even type of columns you use may need to be customized. Your mileage may vary and you are encouraged to experiment and test before deploying to production.
```
Comment (Buzzardo) [inline docs/src/docs/asciidoc/guides/how-to-jpa.adoc:19]: Change to:

NOTE: Except for token, state, metadata, settings, and claims values, we use the JPA default column length of 255 for all columns. In reality, the length and even type of columns you use may need to be customized. You are encouraged to experiment and test before deploying to production.

(I know you didn't change this, but I just noticed it. Sorry.)

File: docs/src/docs/asciidoc/guides/how-to-jpa.adoc [hunk 6]
```diff
@@ -9,19 +9,19 @@
...
 
 [[jpa-client-schema]]
 === Client Schema
 
-The xref:{docs-dir}/core-components.adoc#registered-client-repository[`RegisteredClient`] domain object contains a few multi-valued fields and some settings fields that require storing arbitrary key/value data. The following is an example that we will use to create a JPA entity.
+The xref:{docs-dir}/core-model-components.adoc#registered-client[`RegisteredClient`] domain object contains a few multi-valued fields and some settings fields that require storing arbitrary key/value data. The following is an example that we will use to create a JPA entity.
```
Comment (Buzzardo) [inline docs/src/docs/asciidoc/guides/how-to-jpa.adoc:24]: Change "The following is an example that we will use to create a JPA entity." to "The following example creates a JPA entity."

File: docs/src/docs/asciidoc/guides/how-to-jpa.adoc [hunk 7]
```diff
@@ -32,7 +32,7 @@ include::{examples-dir}/src/main/resources/oauth2-registered-client-schema.sql[]
 [[jpa-authorization-schema]]
 === Authorization Schema
 
-The xref:{docs-dir}/core-components.adoc#oauth2-authorization-service[`OAuth2Authorization`] domain object is more complex, and contains several multi-valued fields as well as numerous arbitrarily long token values, metadata, settings and claims values. The built-in JDBC implementation utilizes a flattened structure that prefers performance over normalization, which we adopt here as well.
+The xref:{docs-dir}/core-model-components.adoc#oauth2-authorization[`OAuth2Authorization`] domain object is more complex, and contains several multi-valued fields as well as numerous arbitrarily long token values, metadata, settings and claims values. The built-in JDBC implementation utilizes a flattened structure that prefers performance over normalization, which we adopt here as well.
```
Comment (Buzzardo) [inline docs/src/docs/asciidoc/guides/how-to-jpa.adoc:35]: Remove the comma in "complex, and contains" (lists that have only two items do not get commas).

File: docs/src/docs/asciidoc/guides/how-to-jpa.adoc [hunk 8]
```diff
@@ -47,7 +47,7 @@ include::{examples-dir}/src/main/resources/oauth2-authorization-schema.sql[]
 [[jpa-authorization-consent-schema]]
 === Authorization Consent Schema
 
-The xref:{docs-dir}/core-components.adoc#oauth2-authorization-consent-service[`OAuth2AuthorizationConsent`] domain object is the simplest to model, and only contains a single multi-valued field in addition to a composite key.  The following is an example that we will use to create a JPA entity.
+The xref:{docs-dir}/core-model-components.adoc#oauth2-authorization-consent[`OAuth2AuthorizationConsent`] domain object is the simplest to model, and only contains a single multi-valued field in addition to a composite key.  The following is an example that we will use to create a JPA entity.
```
Comment (Buzzardo) [inline docs/src/docs/asciidoc/guides/how-to-jpa.adoc:50]: Change "to model, and only contains a single" to "to model and contains only a single" (that is, remove the command and move "only").

File: docs/src/docs/asciidoc/guides/how-to-jpa.adoc [hunk 9]
```diff
@@ -65,7 +65,7 @@ NOTE: The following entities are minimally annotated and are just examples. They
 [[jpa-client-entity]]
 === Client Entity
 
-The following is an example of the `Client` entity which is used to persist information mapped from the xref:{docs-dir}/core-components.adoc#registered-client-repository[`RegisteredClient`] domain object.
+The following is an example of the `Client` entity which is used to persist information mapped from the xref:{docs-dir}/core-model-components.adoc#registered-client[`RegisteredClient`] domain object.
```
Comment (Buzzardo) [inline docs/src/docs/asciidoc/guides/how-to-jpa.adoc:68]: Change to:

The following listing shows the `Client` entity, which is used to persist information mapped from the xref:{docs-dir}/core-model-components.adoc#registered-client[`RegisteredClient`] domain object.

File: docs/src/docs/asciidoc/guides/how-to-jpa.adoc [hunk 10]
```diff
@@ -76,7 +76,7 @@ include::{examples-dir}/src/main/java/sample/jpa/Client.java[tag=class]
 [[jpa-authorization-entity]]
 === Authorization Entity
 
-The following is an example of the `Authorization` entity which is used to persist information mapped from the xref:{docs-dir}/core-components.adoc#oauth2-authorization-service[`OAuth2Authorization`] domain object.
+The following is an example of the `Authorization` entity which is used to persist information mapped from the xref:{docs-dir}/core-model-components.adoc#oauth2-authorization[`OAuth2Authorization`] domain object.
```
Comment (Buzzardo) [inline docs/src/docs/asciidoc/guides/how-to-jpa.adoc:79]: Change to:

The following listing shows the `Authorization` entity, which is used to persist information mapped from the xref:{docs-dir}/core-model-components.adoc#oauth2-authorization[`OAuth2Authorization`] domain object.

File: docs/src/docs/asciidoc/guides/how-to-jpa.adoc [hunk 11]
```diff
@@ -87,7 +87,7 @@ include::{examples-dir}/src/main/java/sample/jpa/Authorization.java[tag=class]
 [[jpa-authorization-consent-entity]]
 === Authorization Consent Entity
 
-The following is an example of the `AuthorizationConsent` entity which is used to persist information mapped from the xref:{docs-dir}/core-components.adoc#oauth2-authorization-consent-service[`OAuth2AuthorizationConsent`] domain object.
+The following is an example of the `AuthorizationConsent` entity which is used to persist information mapped from the xref:{docs-dir}/core-model-components.adoc#oauth2-authorization-consent[`OAuth2AuthorizationConsent`] domain object.
```
Comment (Buzzardo) [inline docs/src/docs/asciidoc/guides/how-to-jpa.adoc:90]: Change to:

The following listing shows the `AuthorizationConsent` entity, which is used to persist information mapped from the xref:{docs-dir}/core-model-components.adoc#oauth2-authorization-consent[`OAuth2AuthorizationConsent`] domain object.

File: docs/src/docs/asciidoc/guides/how-to-jpa.adoc [hunk 12]
```diff
@@ -143,7 +143,7 @@ CAUTION: Keep in mind that writing JSON data to text columns with a fixed length
 [[jpa-registered-client-repository]]
 === Registered Client Repository
 
-The following is an example of the `JpaRegisteredClientRepository` which uses a <<jpa-client-repository,`ClientRepository`>> for persisting a <<jpa-client-entity,`Client`>>, and maps to/from the xref:{docs-dir}/core-components.adoc#registered-client-repository[`RegisteredClient`] domain object.
+The following is an example of the `JpaRegisteredClientRepository` which uses a <<jpa-client-repository,`ClientRepository`>> for persisting a <<jpa-client-entity,`Client`>>, and maps to/from the xref:{docs-dir}/core-model-components.adoc#registered-client[`RegisteredClient`] domain object.
```
Comment (Buzzardo) [inline docs/src/docs/asciidoc/guides/how-to-jpa.adoc:146]: Change to:

The following listing shows the `JpaRegisteredClientRepository`, which uses a <<jpa-client-repository,`ClientRepository`>> for persisting a <<jpa-client-entity,`Client`>> and maps to and from the xref:{docs-dir}/core-model-components.adoc#registered-client[`RegisteredClient`] domain object.

File: docs/src/docs/asciidoc/guides/how-to-jpa.adoc [hunk 13]
```diff
@@ -154,7 +154,7 @@ include::{examples-dir}/src/main/java/sample/jpa/JpaRegisteredClientRepository.j
 [[jpa-authorization-service]]
 === Authorization Service
 
-The following is an example of the `JpaOAuth2AuthorizationService` which uses an <<jpa-authorization-repository,`AuthorizationRepository`>> for persisting an <<jpa-authorization-entity,`Authorization`>>, and maps to/from the xref:{docs-dir}/core-components.adoc#oauth2-authorization-service[`OAuth2Authorization`] domain object.
+The following is an example of the `JpaOAuth2AuthorizationService` which uses an <<jpa-authorization-repository,`AuthorizationRepository`>> for persisting an <<jpa-authorization-entity,`Authorization`>>, and maps to/from the xref:{docs-dir}/core-model-components.adoc#oauth2-authorization[`OAuth2Authorization`] domain object.
```
Comment (Buzzardo) [inline docs/src/docs/asciidoc/guides/how-to-jpa.adoc:157]: Change to:

The following listing shows the `JpaOAuth2AuthorizationService`, which uses an <<jpa-authorization-repository,`AuthorizationRepository`>> for persisting an <<jpa-authorization-entity,`Authorization`>> and maps to and from the xref:{docs-dir}/core-model-components.adoc#oauth2-authorization[`OAuth2Authorization`] domain object.

File: docs/src/docs/asciidoc/guides/how-to-jpa.adoc [hunk 14]
```diff
@@ -165,10 +165,10 @@ include::{examples-dir}/src/main/java/sample/jpa/JpaOAuth2AuthorizationService.j
 [[jpa-authorization-consent-service]]
 === Authorization Consent Service
 
-The following is an example of the `JpaOAuth2AuthorizationConsentService` which uses an <<jpa-authorization-consent-repository,`AuthorizationConsentRepository`>> for persisting an <<jpa-authorization-consent-entity,`AuthorizationConsent`>>, and maps to/from the xref:{docs-dir}/core-components.adoc#oauth2-authorization-consent-service[`OAuth2AuthorizationConsent`] domain object.
+The following is an example of the `JpaOAuth2AuthorizationConsentService` which uses an <<jpa-authorization-consent-repository,`AuthorizationConsentRepository`>> for persisting an <<jpa-authorization-consent-entity,`AuthorizationConsent`>>, and maps to/from the xref:{docs-dir}/core-model-components.adoc#oauth2-authorization-consent[`OAuth2AuthorizationConsent`] domain object.
```
Comment (Buzzardo) [inline docs/src/docs/asciidoc/guides/how-to-jpa.adoc:168]: Change to:

The following listing shows the `JpaOAuth2AuthorizationConsentService`, which uses an <<jpa-authorization-consent-repository,`AuthorizationConsentRepository`>> for persisting an <<jpa-authorization-consent-entity,`AuthorizationConsent`>> and maps to and from the xref:{docs-dir}/core-model-components.adoc#oauth2-authorization-consent[`OAuth2AuthorizationConsent`] domain object.

**Review by Buzzardo — CHANGES_REQUESTED** [review]

**Review by sjohnr — COMMENTED** [review]
This is great, @jgrandja! I've added a couple of questions that could lead to improvements.

@Buzzardo, I can take the comments on the `how-to-jpa.adoc` and put them in a separate PR if @jgrandja agrees.

**Comment by Buzzardo** [comment]
> @Buzzardo, I can take the comments on the `how-to-jpa.adoc` and put them in a separate PR if @jgrandja agrees.

Fine by me, so long as it works for you folks.

File: docs/src/docs/asciidoc/core-model-components.adoc [hunk 15]
```diff
@@ -0,0 +1,470 @@
...
+	private TokenSettings tokenSettings;    <12>
+
+	...
+
+}
+----
+<1> `id`: The ID that uniquely identifies the `RegisteredClient`.
+<2> `clientId`: The client identifier.
+<3> `clientIdIssuedAt`: The time at which the client identifier was issued.
+<4> `clientSecret`: The client’s secret.
```
[the remark this answers is not in the document]
Comment (jgrandja) [reply docs/src/docs/asciidoc/core-model-components.adoc:85]: Good catch. I added a reference link

**Pull request closed without merging**